### PR TITLE
Fix the dialog behavior to act more like a standard Material dialog - Add dialog to demo

### PR DESCRIPTION
**Pull request by jneidlinger**
This fixes some of the dialog behavior to behave more in-line with how material dialogs work.
When shown, they should shadow anything in the background, and center in the UI. Clicking any area outside of the dialog effectively dismisses the dialog. Added the ability to make the dialog have no confirmation button options either, for use with simple radio-button type option dialogs.

Included is an example in the demo project.

image

image

Also I fixed up a WIP client side decoration that was showing no matter what.

### CODE REVIEW

**Comment by ricardomv** [comment]
@jneidlinger in the simple dialog i think the title should not hide so you should remove it from the flickable, or maybe even remove the flickable an show all the radio buttons

**Comment by ricardomv** [comment]
also there is a big blank space in the bottom of the simple dialog

**Comment by ricardomv** [comment]
can you make the demo only open in the demo area not over the hole application. simple menus should be able to work without obscuring the app

**Comment by ricardomv** [comment]
in adition to clicking outside of the window it would be nice to press 'esc' to close the dialog

**Comment by jneidlinger** [comment]
@ricardomv Dialogs should obscure the whole app. The dialog should be dismissed, before any interaction continues in the application. This follows googles guidelines for dialogs:

http://www.google.com/design/spec/components/dialogs.html#

Also, the buttons should not be raised in dialogs.

**Comment by ricardomv** [comment]
@jneidlinger not in the dialog in the demo. that's were i commented

**Comment by ricardomv** [comment]
@jneidlinger i wasss talking about this case 
image

**Comment by jneidlinger** [comment]
The dropdowns still function without obscuring the app, only dialogs obscure the app.

**Comment by ricardomv** [comment]
it used to be the other way around the dropdown was fullscreen and the dialog could be placed by the client (programmer)
```
    Component.onCompleted: {
        parent = Utils.findRootChild(view, "overlayLayer")
        parent.shadowBackground = true
    }
```
i just wouldn't like to see this all over the code so either remove it in the dropdown or in the dialog or merge the both.

**Comment by ricardomv** [comment]
it would be nice when you add the demo to have a separate commit for it when you do that either use commit --amend for the tweaks or now you can use interactive rebase

**Comment by jneidlinger** [comment]
The code you are noting is only in components that use a overlayLayer, currently the dropdown, and the dialog, so it wouldn't be "all over" the code. The dropdown component is relatively new, so I'm not sure what you are referring to.

**Comment by ricardomv** [comment]
is it possible to make a dialog using the Dropdown then?

**Comment by jneidlinger** [comment]
I see, it may be possible to base the dialog on the dropdown code. Let me look into it.

**Comment by ricardomv** [comment]
@jneidlinger don't create a new PR reset this one and start from there

**Comment by iBeliever** [comment]
@jneidlinger It might be easiest to split the base code out of Dropdown and into a PopoverBase class, for use by both Dialog and Dropdown

@ricardomv When you comment on code in a pull request, it's generally easiest to read if you do so on the overall diff instead of a specific commit, because then the surrounding code shows in the comment thread.

**Comment by jneidlinger** [comment]
@iBeliever Yes, I'm pulling the logic into a base component to be shared. I'll re-open this pull request when its ready.

**Comment by ricardomv** [comment]
@iBeliever thanks i didn't know that, i thought github just changed the way the commentss are displyed here.
@jneidlinger sounds good :+1:

**Comment by jneidlinger** [comment]
Cleaned up the dialog to behave a little better.

Dialog and Dropdown are now based on a common Popover component.

**Comment by ricardomv** [comment]
looks much better, but i still have a few issues, maybe we should rename Popover to Dropdown because currently the dropdown only holds the transitions that can be moved to the currently named Popover type. About the transitions i don't really like the width height animations, the contents of the dialog appear from nowhere and it looks a bit messy.

**Comment by jneidlinger** [comment]
Open to the animations changing, was trying to follow how dialogs open on Android. Maybe the width and height shouldn't start at 0 for dialogs?

**Comment by jneidlinger** [comment]
The reason Dropdown has its own transitions defined is so that it, and Dialog can have differing open/close transitions.

**Comment by ricardomv** [comment]
@jneidlinger i don't know what transitions to use maybe @iBeliever has an opinion

File: modules/Material/Dialog.qml [hunk 1]
```diff
@@ -19,118 +19,186 @@
...
+
+    default property alias dialogContent: placeholder.children
+
+    signal accepted()
+    signal rejected()
+
+    transitions: [
+        Transition {
+            from: "open"
+            to: "closed"
```
Comment (ricardomv) [inline modules/Material/Dialog.qml]: since these transitions are the same as closed->open you can remove these two lines and the next Transition

File: modules/Material/Dialog.qml [hunk 2]
```diff
@@ -19,118 +19,186 @@
...
+            anchors.fill: parent
+            propagateComposedEvents: false
+            onClicked: {
+                mouse.accepted = false
+            }
+        }
+    }
+
+    function show() {
+        open(Utils.findRoot(parent))
```
Comment (ricardomv) [inline modules/Material/Dialog.qml]: i haven't tested this but doesn't this work with just parent since you are setting it in Popover?

File: modules/Material/Toolbar.qml [hunk 3]
```diff
@@ -153,7 +153,7 @@ View {
 
     Row {
         id: windowControls
-
+        visible: clientSideDecorations
```
Comment (ricardomv) [inline modules/Material/Toolbar.qml]: again this has nothing to do with any of the commits

Comment (jneidlinger) [inline modules/Material/Toolbar.qml]: I mentioned it in the pull request. If we are going to be picky and force a separate pull request / branch for a tiny tiny fix like this, than so be it. That being said, I didn't think its a big deal to have this fixed here.

File: demo/demo.qmlproject [hunk 4]
```diff
@@ -16,5 +16,5 @@ Project {
         directory: "."
     }
 
-    importPaths: [ "../modules" ]
+    importPaths: [ "../modules", "../modules/Material", "../../qmlextras/modules", "../../qmlextras/modules/Material"]
```
Comment (iBeliever) [inline demo/demo.qmlproject]: Why have `../modules/Material` and `../../qmlextras/modules/Material`? I thought you only needed to import the directory where modules exist, not the module directories themselves.

Comment (jneidlinger) [inline demo/demo.qmlproject]: Oops, sorry, that wasn't supposed to be committed... I was testing something.

**Comment by ricardomv** [comment]
@jneidlinger please squash the commits with git rebase
and push with git push --force
https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase-i/

File: modules/Material/ApplicationWindow.qml [hunk 5]
```diff
@@ -111,12 +111,39 @@ Controls.ApplicationWindow {
         z: 100
 
         property Item currentOverlay
+        property bool backdrop: false
+
+        Rectangle {
```
Comment (iBeliever) [inline modules/Material/ApplicationWindow.qml]: How about instead of having the child rectangle, we make `overlayLayer` be a Rectangle, and set the color to something like `currentOverlay.overlayColor`? This would allow for future components requiring move background colors in the future.

File: modules/Material/Dialog.qml (hunk 1 above)
Comment (iBeliever) [inline modules/Material/Dialog.qml]: You may need to use `from: "*"` and `to: "*"`, which is what I've seen in examples. I'm not sure if you can just remove them completely.

File: modules/Material/ApplicationWindow.qml (hunk 5 above)
Comment (jneidlinger) [inline modules/Material/ApplicationWindow.qml]: Good idea, I'll do that.

File: modules/Material/Toolbar.qml (hunk 3 above)
Comment (iBeliever) [inline modules/Material/Toolbar.qml]: I wouldn't worry about it being unrelated to the rest of the pull request. Code reviews are supposed to improve code quality, not slow us down by forcing us to go through extra work and processes.

Comment (ricardomv) [inline modules/Material/Toolbar.qml]: i don't mean a new PR sorry but i would put it in a new commit

File: modules/Material/Dialog.qml (hunk 1 above)
Comment (ricardomv) [inline modules/Material/Dialog.qml]: i tested without it and it worked fine

File: modules/Material/Toolbar.qml (hunk 3 above)
Comment (iBeliever) [inline modules/Material/Toolbar.qml]: Yes, a new commit would ideally be best, so in the unlikely event we could easily revert a speicific commit without reverting unrelated behavior.

**Comment by iBeliever** [comment]
Regarding the animations, I'm not entirely sure what Android does (or if it's animated at all in the current version), and the Google spec does't say. How about we keep the height and width the same, but animate the opacity and have the dialog slide in/out from the top of the window?

**Comment by jneidlinger** [comment]
I'm going to tone down the sizing, and just animate opacity for now. We can also circle back later and allow for custom open / close animations when defining a new dialog. That is what all the material web frameworks do.

**Comment by jneidlinger** [comment]
@ricardomv I squashed the commits, and made sure to leave the simple client side decorations fix in its own commit.

**Comment by iBeliever** [comment]
Great work, thanks!
